update_db.py

```diff
@@ -0,0 +1,197 @@
+import base64
+import json
+import re
+from apiclient import errors
+from collections import namedtuple
+
+import app
+from app.lib.gmail_api import GmailApi
+from app.models import User, Message
+from config.secrets import USER_INFO_BY_EMAIL, GMAIL_QUERIES
+
+
+# TODO don't run query up here?
+EMAIL_TO_USER_ID = {user.email: user.id for user in User.query.all()}
+
+db = app.db
+service = GmailApi().get_service()
+
+MessageData = namedtuple('MessageData', [
+    'body',
+    'timestamp',
+    'message_id',
+    'label',
+    'data',
+    'sender_user_id',
+    'thread_id',
+])
+
+
+def create_or_update_users():
+    for email, user_attributes in USER_INFO_BY_EMAIL.items():
+        existing_user = User.query.filter_by(email=email).first()
+        if existing_user is not None:
+            print("Updating: {}".format(email))
+            User.query.filter_by(email=email).update(user_attributes)
+        else:
+            print("Creating: {}".format(email))
+            new_user = User(**user_attributes)
+            db.session.add(new_user)
+
+    db.session.commit()
+
+
+def fetch_and_save_new_messages(query):
+    """
+    Get all messages from gmail api and save to db. This happens in a few steps - 
+    1. use gmail api messages.list() to get a list of all message ids returned from query
+    2. loop through each of those messages ids and use messages.get() to get more detailed info
+    3. parse through response so we can get get the fields we want to save in the db
+
+    Params
+        query [String] passed to gmailAPI to fetch messages
+    Returns
+        [Int] Count of new messages saved
+    """
+    existing_message_ids = set(m.message_id for m in Message.query.all())
+
+    try:
+        print("Fetching new messages from query: {}".format(query))
+        response = service.users().messages().list(userId='me', q=query).execute()
+        messages = response.get('messages', [])
+
+        while 'nextPageToken' in response:
+            response = service.users().messages().list(
+                userId='me',
+                q=query,
+                pageToken=response['nextPageToken']
+            ).execute()
+            messages.extend(response.get('messages', []))
+
+    except errors.HttpError as error:
+        print('An error occurred retrieving mesages: {}'.format(error))
+
+    new_message_ids = [m['id'] for m in messages if m['id'] not in existing_message_ids]
+
+    print("Found {} new messages".format(len(new_message_ids)))
+    for count, message_id in enumerate(new_message_ids):
```

Comment:
Instead of making a new HTTP request for every message id, we could try to make batched requests (from https://developers.google.com/api-client-library/python/guide/batch):

```
from apiclient.http import BatchHttpRequest
CHUNK_SIZE = 100
def _chunk(l, size):
    for i in range(0, len(l), size):
        yield l[i:i+size]

def _parse_response_and_save(response):
   parsed_response = _parse_response(response)
   new_message = Message(...)
   db.session.add(new_message)

chunks = _chunk(message_ids, CHUNK_SIZE)
for chunk in chunks:
    batch = service.new_batch_http_request(callback=_parse_response_and_save)
    for message_id in chunk:
        batch.add(service.users().messages().get(userId = 'me', id = message_id))
    batch.execute()
    db.session.commit()
    count += len(chunk)
    num_messages_left = len(new_message_ids) - count
    print('{} messages to go'.format(num_messages_left))
```